Avoid catastrophic backtracing in regex

Ravlen opened the pull request:
@tancnle This PR attempts to fix some regex that could cause catastrophic backtracing, could you take a look?

## For the `ghc_core` lexer:

- <img width="755" alt="Screen Shot 2021-03-10 at 15 57 39" src="https://user-images.githubusercontent.com/36979740/110595228-1d966980-81c1-11eb-93f7-d12ecc11c393.png">

- <img width="755" alt="Screen Shot 2021-03-10 at 15 58 13" src="https://user-images.githubusercontent.com/36979740/110595251-22f3b400-81c1-11eb-8020-471ea56c4377.png">

## For the `factor` lexer:

- <img width="755" alt="Screen Shot 2021-03-10 at 16 09 54" src="https://user-images.githubusercontent.com/36979740/110595274-2be48580-81c1-11eb-82f3-d48039acc697.png">

- <img width="755" alt="Screen Shot 2021-03-10 at 16 09 39" src="https://user-images.githubusercontent.com/36979740/110595299-33a42a00-81c1-11eb-9b02-9cc2988cb0ae.png">

Ravlen commented on lib/rouge/lexers/factor.rb:
```diff
@@ -243,7 +243,7 @@ def self.builtins
 
         # strings
-        rule %r/"""\s+.*?\s+"""/, Str
+        rule %r/"""[^"].*?[^"]"""/, Str
```
Actually, I think this is wrong, let me fix that.

Ravlen replied:
Maybe it's not even supported: https://github.com/factor/factor/blob/master/core/strings/strings-tests.factor

tancnle replied:
Interesting. Looks like GitHub is parsing the code correctly'ish 🤔 The highlight for double quotes seems to be out of place but overall it looks good.

I was suggesting lifting the regexes from `Pygments`. However, it seems Pygments 2.8.1 (latest), the Python package that Rouge compatible with does not handle triple quotes well (could be an existing bug).

<img width="736" alt="Screen Shot 2021-03-11 at 8 17 00 pm" src="https://user-images.githubusercontent.com/756722/110764468-36794a80-82a7-11eb-852e-f693529fa562.png">

tancnle replied:
According to https://github.com/factor/factor/issues/1474, it seems like triple quotes are not supported.

tancnle commented:
@Ravlen What do you think about getting the regexes from [Pygments](https://github.com/pygments/pygments/blob/master/pygments/lexers/factor.py#L267-L269)?

```diff
diff --git lib/rouge/lexers/factor.rb lib/rouge/lexers/factor.rb
index 2a18f715..4a7458a5 100644
--- lib/rouge/lexers/factor.rb
+++ lib/rouge/lexers/factor.rb
@@ -243,8 +243,9 @@ module Rouge
         end
 
         # strings
-        rule %r/"""\s+.*?\s+"""/, Str
-        rule %r/"(\\.|[^\\])*?"/, Str
+        rule %r/"""\s(?:.|\n)*?\s"""/, Str
+        rule %r/"(?:\\\\|\\"|[^"])*"/, Str
+        rule %r/\S+"\s+(?:\\\\|\\"|[^"])*"/, Str
         rule %r/(CHAR:)(\s+)(\\[\\abfnrstv]*|\S)(?=\s)/, Str::Char
 
         # comments
```

dblessing commented:
@tancnle I like this idea. They're vetted/tested.

Ravlen commented:
@tancnle @dblessing OK, updated with the regex from Pygments, but I dropped the `"""` string regex as I'm quite confident it isn't supported by Factor. See https://github.com/factor/factor/blob/master/core/strings/strings-tests.factor, and the commen Tan found at: https://github.com/factor/factor/issues/1474#issuecomment-409751485

tancnle commented:
Travis CI seems to be problematic 🤔 @dblessing could you help to check what is going wrong with Travis integration?

dblessing commented:
It looks like we've run out of build minutes due to Travis' new billing - https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing. 

We either need to apply for OSS allotment, or potentially use GitLab builds. What do you all think @tancnle @Ravlen. I think GitLab only offers 400 free minutes, but maybe they'd be willing to sponsor a 'plan' for Rouge?

tancnle commented:
@dblessing Would it be much hassle to apply for an OSS allotment? It provides 10k credits which are build minutes I believe.

I think [BuildKite](https://buildkite.com/pricing) or [CircleCI](https://circleci.com/open-source/) might be a better choice given they have no build minute limit for OSS.

dblessing commented:
@tancnle I've applied for credits. But it's not 10,000 automatic - I had to request the amount I think we need. I asked if they had visibility into historical usage. But AFAICT we use about 10 minutes per build - 1 minute per suite and we test 10 Ruby versions. Probably just 500-1000 minutes will suffice.

tancnle commented:
@dblessing I think moving to BuildKite or CircleCI with unlimited build minute will remove the need to recalculate the allotment as we scale our pipeline.

>  we test 10 Ruby versions

That's a lot. Are we thinking about deprecating old Ruby versions (I.e. no ongoing supported for MRI < 2.5)?

dblessing commented:
Is honestly prefer either to stay with Travis or move to GitLab.

I agree we should consider removing some Ruby version support.

On Mon, Mar 22, 2021 at 7:20 PM Tan Le ***@***.***> wrote:

> @dblessing <https://github.com/dblessing> I think moving to BuildKite or
> CircleCI with unlimited build minute will remove the need to recalculate
> the allotment as we scale our pipeline.
>
> we test 10 Ruby versions
>
> That's a lot. Are we thinking about deprecating old Ruby versions (I.e. no
> ongoing supported for MRI < 2.5)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rouge-ruby/rouge/pull/1690#issuecomment-804489087>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASIVFAEBMWFG7F5KHLU7ATTE7NE5ANCNFSM4Y5NLKIQ>
> .
>
-- 
Drew

tancnle commented:
No worries @dblessing. I am cool with staying with TravisCI and get this MR unblocked and merged. FYI I have created the [issue](https://github.com/rouge-ruby/rouge/issues/1694) to remove support for old Ruby version.

Ravlen commented:
@dblessing @tancnle If we want to switch to GitLab, I'm reasonably confident I could write the pipeline configuration. There are 50k minutes available for open source projects, see https://about.gitlab.com/solutions/open-source/, if we want to look into it (though perhaps you'd need to move the whole project to GitLab to get the full 50k?)

dblessing commented:
I still haven't heard back from Travis so I pinged them again on the support issue. 

@Ravlen I think you're right - we'd need to either move the whole project or see if they would grant us a Premium license. I think the best case is Travis comes through for us, but if not we can see about GitLab CI.

tancnle commented:
@dblessing Any news from Travis? If not, I think we can start migrating to other CI solution.

dblessing commented:
@tancnle I thought I commented about 3 weeks ago saying we had CI
allocation and asking @Ravlan to push again. Either a malfunction or I only
dreamed it. Sorry all. It should be working now, though.

On Tue, May 18, 2021 at 7:06 PM Tan Le ***@***.***> wrote:

> @dblessing <https://github.com/dblessing> Any news from Travis? If not, I
> think we can start migrating to another CI solution.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rouge-ruby/rouge/pull/1690#issuecomment-843646595>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASIVFFMSNFZO6EJVGW3FJTTOL6J3ANCNFSM4Y5NLKIQ>
> .
>
-- 
Drew

Ravlen commented:
@tancnle @dblessing Yay, finally the Travis build worked 😁

jneen commented:
I for one would be 100% down to move to gitlab, so long as we could preserve the state of the issue tracker.

jneen commented:
(And given that this software is used *by* gitlab I think it would be fitting :] )
